src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+    }
+
+    switch (res)
+    {
+    case Result::kDrop:
+        break;
+    case Result::kReplyIcmp:
+        // TODO: Implement the logic for replying ICMP packets.
+        break;
+    case Result::kForward:
+        ip6Header.SetPayloadLength(aMessage.GetLength() - aMessage.GetOffset());
+        Checksum::UpdateMessageChecksum(aMessage, ip6Header.GetSource(), ip6Header.GetDestination(),
+                                        ip6Header.GetNextHeader());
+        err = aMessage.PrependBytes(&ip6Header, sizeof(ip6Header));
+        if (err != kErrorNone)
+        {
+            // This might happen when the platform failed to reserve enough space before the original IPv4 packet.
+            LogWarn("failed to prepend IPv6 head to translated message");
+            res = Result::kDrop;
+        }
+        break;
+    }
+
+exit:
+    return res;
+}
+
+void Nat64Translator::ReleaseMapping(AddressMapping &aMapping)
+{
+    LogInfo("mapping removed: %s -> %s", aMapping.mIp6.ToString().AsCString(), aMapping.mIp4.ToString().AsCString());
+    mIp4AddressPool[mAvailableAddressCount] = aMapping.mIp4;
+    mAvailableAddressCount++;
+    mAddressMappingPool.Free(aMapping);
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::CreateMapping(const Ip6::Address &aAddr)
+{
+    AddressMapping *mapping = mAddressMappingPool.Allocate();
+
+    if (mapping == nullptr)
+    {
+        uint64_t                   now = Get<Uptime>().GetUptime();
+        LinkedList<AddressMapping> idleMappings;
+
+        mActiveAddressMappings.RemoveAllMatching(now, idleMappings);
+
+        for (AddressMapping *idleMapping = idleMappings.Pop(); idleMapping != nullptr; idleMapping = idleMappings.Pop())
+        {
+            ReleaseMapping(*idleMapping);
+        }
+
+        mapping = mAddressMappingPool.Allocate();
+    }
+
+    if (mAvailableAddressCount == 0)
+    {
+        if (mapping != nullptr)
+        {
+            mAddressMappingPool.Free(*mapping);
+            mapping = nullptr;
+        }
+        ExitNow();
+    }
+
+    VerifyOrExit(mapping != nullptr);
+
+    mapping->mIp6 = aAddr;
+    mAvailableAddressCount--;
+    mapping->mIp4 = mIp4AddressPool[mAvailableAddressCount];
+    mapping->Touch(Get<Uptime>().GetUptime());
+    mActiveAddressMappings.Push(*mapping);
+    LogInfo("mapping created: %s -> %s", mapping->mIp6.ToString().AsCString(), mapping->mIp4.ToString().AsCString());
+
+exit:
+    return mapping;
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::GetMapping(const Ip6::Address &aAddr, bool aTryCreate)
+{
+    AddressMapping *mapping = mActiveAddressMappings.FindMatching(aAddr);
+
+    // Exit if we found a valid mapping.
+    VerifyOrExit(mapping == nullptr);
+    // If we don't have a valid mapping and we don't want to create one.
+    VerifyOrExit(aTryCreate);
+
+    mapping = CreateMapping(aAddr);
+
+exit:
+    return mapping;
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::GetMapping(const Ip4::Address &aAddr)
```

Comment:
Use `aIp4Address` for param name to make it clear it is `Ip4::Address`?

```suggestion
Nat64Translator::AddressMapping *Nat64Translator::GetMapping(const Ip4::Address &aIp4Address)
```